internal/controller/connectionsecret/requestname_extractor.go

```diff
@@ -0,0 +1,292 @@
+	var ids ConnSecretIdentifiers
+
+	// === Internal format: <ProjectID>$<ClusterName>$<DatabaseUserName>
+	if strings.Contains(req.Name, InternalSeparator) {
+		parts := strings.SplitN(req.Name, InternalSeparator, 3)
+		if len(parts) != 3 {
+			return ids, ErrInternalFormatPartsInvalid
+		}
+		if parts[0] == "" || parts[1] == "" || parts[2] == "" {
+			return ids, ErrInternalFormatPartEmpty
+		}
+		return ConnSecretIdentifiers{
+			ProjectID:        parts[0],
+			ClusterName:      parts[1],
+			DatabaseUsername: parts[2],
+		}, nil
+	}
+
+	// === K8s format: <ProjectName>-<ClusterName>-<DatabaseUserName>
+	var secret corev1.Secret
+	if err := c.Get(ctx, req, &secret); err != nil {
+		return ids, err
+	}
+
+	labels := secret.GetLabels()
+	projectID, hasProject := labels[ProjectLabelKey]
+	clusterName, hasCluster := labels[ClusterLabelKey]
+
+	// Missing labels or values
+	if !hasProject || !hasCluster {
+		return ids, ErrK8sLabelsMissing
+	}
+	if projectID == "" || clusterName == "" {
+		return ids, ErrK8sLabelEmpty
+	}
+
+	sep := fmt.Sprintf("-%s-", clusterName)
+	parts := strings.SplitN(req.Name, sep, 2)
+	if len(parts) != 2 {
+		return ids, ErrK8sNameSplitInvalid
+	}
+	if parts[0] == "" || parts[1] == "" {
+		return ids, ErrK8sNameSplitEmpty
+	}
+
+	return ConnSecretIdentifiers{
+		ProjectID:        projectID,
+		ProjectName:      parts[0],
+		ClusterName:      clusterName,
+		DatabaseUsername: parts[1],
+	}, nil
+}
+
+// LoadPairedResources fetches the paired AtlasDeployment and AtlasDatabaseUser forming the ConnectionSecret
+// using the registered indexers
+func LoadPairedResources(ctx context.Context, c client.Client, ids ConnSecretIdentifiers, namespace string) (*ConnSecretPair, error) {
+	compositeDeploymentKey := ids.ProjectID + "-" + ids.ClusterName
+	deployments := &akov2.AtlasDeploymentList{}
+	if err := c.List(ctx, deployments, &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(indexer.AtlasDeploymentBySpecNameAndProjectID, compositeDeploymentKey),
+	}); err != nil {
+		return nil, err
+	}
+
+	compositeUserKey := ids.ProjectID + "-" + ids.DatabaseUsername
+	users := &akov2.AtlasDatabaseUserList{}
+	if err := c.List(ctx, users, &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(indexer.AtlasDatabaseUserBySpecUsernameAndProjectID, compositeUserKey),
+	}); err != nil {
+		return nil, err
+	}
+
+	switch {
+	case len(deployments.Items) == 0 && len(users.Items) == 0:
+		return nil, ErrNoPairedResourcesFound
+	case len(deployments.Items) == 0:
+		return &ConnSecretPair{
+			ProjectID:  ids.ProjectID,
+			Deployment: nil,
+			User:       &users.Items[0],
+		}, ErrNoDeploymentFound
+	case len(users.Items) == 0:
+		return &ConnSecretPair{
+			ProjectID:  ids.ProjectID,
+			Deployment: &deployments.Items[0],
+			User:       nil,
+		}, ErrNoUserFound
+	case len(deployments.Items) > 1:
+		return nil, ErrManyDeployments
+	case len(users.Items) > 1:
+		return nil, ErrManyUsers
+	}
```

Comment:
Not need to split every single possible error when they basically mean the same: there is no deploy/user or more than one deploy/user
```suggestion
	if len(deployments.Items) == 0 || len(users.Items) == 0 {
	    return errors.New("deployment or user is missing") // just example of message, adjust it for best
	}
```